### PR TITLE
Add nested mapping support via new `ObjectMap` class

### LINKED ISSUE #63: Add nested mapping
```php
class A {
  private string $var;
}

class B {
  private A $a;
}

$data = [
  'key' => 'test',
];
```

Need to be able to specify that `$data['key']` should be mapped to `$b->a->var`.

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
Fix #63

### CODE REVIEW

File: docs/guide/en/mapping.md [hunk 1]
```diff
@@ -44,6 +44,40 @@ $post = $hydrator->create(Post::class, new ArrayData($data, $map));
...
+    'title' => 'Hello, World!',
+    'textBody' => 'Nice to meet you.',
+    'htmlBody' => '<h1>Nice to meet you.</h1>',
+];
+$map = [
+    'subject' => 'title',
+    'body' => new ObjectMap([
+        'text' => 'textBody',
+        'html' => 'htmlBody',    
+    ]),
```
Comment (samdark) [inline docs/guide/en/mapping.md:76]: How about constructor, method calls, immutable objects etc.?

File: src/ObjectMap.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Hydrator;
+
+use function array_key_exists;
+
+/**
+ * Class provides mapping object property names to keys in the data array.
```
Comment (samdark) [inline src/ObjectMap.php:10]:
```suggestion
 * Class provides a mapping of object property names to keys in the data array.
```

File: src/ObjectMap.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+     *  an object.
+     * @psalm-param MapType $map
+     */
+    public function __construct(
+        public readonly array $map
+    ) {
+    }
+
+    /**
+     * @psalm-return string|list<string>|ObjectMap|null
```
Comment (samdark) [inline src/ObjectMap.php:27]: Need description.

File: src/ObjectMap.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public function getPath(string $name): string|array|self|null
+    {
+        return $this->map[$name] ?? null;
+    }
+
+    /**
+     * @return string[]
```
Comment (samdark) [inline src/ObjectMap.php:35]: Need description.

File: src/ObjectMap.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+     * @psalm-return list<string>
+     */
+    public function getNames(): array
+    {
+        return array_keys($this->map);
+    }
+
+    public function exists(string $name): bool
```
Comment (samdark) [inline src/ObjectMap.php:43]: Need description.

File: src/ObjectMap.php (hunk 4 above)
Comment (vjik) [reply src/ObjectMap.php]: done

File: src/ObjectMap.php (hunk 3 above)
Comment (vjik) [reply src/ObjectMap.php:29]: done

File: docs/guide/en/mapping.md (hunk 1 above)
Comment (vjik) [reply docs/guide/en/mapping.md:76]: It has nothing to do with mapping.

We pass key/value array to hydrator. Create/hydrate logic in `Hydrator`, not in mapping.

File: src/ObjectMap.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Hydrator;
+
+use function array_key_exists;
+
+/**
+ * Class provides a mapping of object property names to keys in the data array.
```
Comment (samdark) [inline src/ObjectMap.php:10]:
```suggestion
 * Provides a mapping of object property names to keys in the data array.
```

File: src/ObjectMap.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     *  an object.
+     * @psalm-param MapType $map
+     */
+    public function __construct(
+        public readonly array $map
+    ) {
+    }
+
+    /**
+     * Returns a path for a given property name or null if mapping is not exist.
```
Comment (samdark) [inline src/ObjectMap.php:27]:
```suggestion
     * Returns a path for a given property name or null if mapping dosen't exist.
```

**Review by samdark — APPROVED** [review]

**Pull request merged**